dali/kernels/algebra/linear_transformation_gpu.h

```diff
@@ -0,0 +1,198 @@
+          ret[j] = ts[i];
+        }
+        return ret;
+    };
+
+    auto get_channels = [](const TensorShape<spatial_ndims + 1> &ts) -> auto {
+        return ts.shape.back();
+    };
+
+    sample.in_size = get_size(in.tensor_shape(i));
+    sample.out_size = get_size(out.tensor_shape(i));
+    sample.in_channels = get_channels(in.tensor_shape(i));
+    sample.out_channels = get_channels(out.tensor_shape(i));
+    sample.in_strides = {sample.in_channels, sample.in_size.x * sample.in_channels};
+    sample.out_strides = {sample.out_channels, sample.out_size.x * sample.out_channels};
+    sample.transformation_matrix = transformation_matrices[i];
+  }
+
+  return ret;
+}
+
+
+}  // namespace detail
+
+
+template <class OutputType, class InputType, size_t N, size_t M, size_t spatial_ndims>
+void __global__ LinearTransformationKernel(
+        const detail::SampleDescriptor<OutputType, InputType, N, M, spatial_ndims> *samples,
+        const BlockDesc<spatial_ndims> *blocks) {
+  const auto &block = blocks[blockIdx.x];
+  const auto &sample = samples[block.sample_idx];
+
+  const Surface2D<const InputType> in = {
+          sample.in, sample.in_size.x, sample.in_size.y, sample.in_channels,
+          sample.in_strides.x, sample.in_strides.y, 1
+  };
+
+  const Surface2D<OutputType> out = {
+          sample.out, sample.out_size.x, sample.out_size.y, sample.out_channels,
+          sample.out_strides.x, sample.out_strides.y, 1
+  };
+
+  for (int y = threadIdx.y + block.start.y; y < block.end.y; y += blockDim.y) {
+    for (int x = threadIdx.x + block.start.x; x < block.end.x; x += blockDim.x) {
+      // Iterate over channels: `o` over output, `i` over input
+      // It's a plain matrix multiplication
+      for (int o = 0; o < N; o++) {
+        OutputType val = 0;
+        for (int i = 0; i < M; i++) {
+          val += sample.transformation_matrix.at(o, i) * in(x, y, i);
+        }
+        out(x, y, o) = val;
+      }
+    }
+  }
+}
+
+
+template <typename OutputType, typename InputType, size_t N, size_t M, size_t spatial_ndims>
+class LinearTransformationGpu {
+ private:
+  static constexpr auto ndims_ = spatial_ndims + 1;
+  using mat = ::dali::mat<N, M, detail::TransformationMatrixElementType>;
+  using BlockDesc = kernels::BlockDesc<spatial_ndims>;
+  using SampleDescriptor = detail::SampleDescriptor<OutputType, InputType, N, M, spatial_ndims>;
+
+  std::vector<SampleDescriptor> sample_descriptors_;
+
+ public:
+  BlockSetup<spatial_ndims,
+          spatial_ndims /* Assumed, that the channel dim is the last dim */> block_setup_;
+
+
+  KernelRequirements
+  Setup(KernelContext &context, const InListGPU<InputType, ndims_> &in,
+        const std::vector<mat> &transformation_matrices,
+        const std::vector<Roi<spatial_ndims>> &rois = {}) {
+    DALI_ENFORCE(rois.empty() || rois.size() == static_cast<size_t>(in.num_samples()),
+                 "Provide ROIs either for all or none input tensors");
+    DALI_ENFORCE([=]() -> bool {
```

Comment:
This is extremely user-unfriendly. You check something, but don't tell the user what the error was, even though you have the offending value at hand. Put the enforce inside a loop over ROIs and throw a meaningful error, with sample index and possibly the offending values.

Reply:
Done